flanker/addresslib/address.py

```diff
@@ -469,30 +645,47 @@ class UrlAddress(Address):
     data", use the parse() and parse_list() functions instead.
     """
 
-    __slots__ = ['address', 'parse_result']
+    _address = None
+    _addr_type = Address.Type.Url
 
-    def __init__(self, spec):
-        self.address = spec
-        self.parse_result = urlparse(spec)
-        self.addr_type = self.Type.Url
+    def __init__(self, raw=None, address=None):
+
+        if raw:
+            if isinstance(raw, unicode):
+                raw = raw.encode('utf-8')
+            parser = url_parser
+            url = parser.parse(raw, lexer=lexer.clone())
+            self._address = url.address.decode('utf-8')
+        elif address:
+            self._address = address
+        else:
+            raise SyntaxError('failed to create UrlAddress: bad parameters')
+
+    @property
+    def address(self):
+        return self._address
+
+    @property
+    def addr_type(self):
+        return self._addr_type
 
     @property
     def hostname(self):
-        hostname = self.parse_result.hostname
+        hostname = urlparse(self._address).hostname
         if hostname:
             return hostname.lower()
 
     @property
     def port(self):
-        return self.parse_result.port
+        return urlparse(self._address).port
```

Comment:
It does not seem right to call `urlparse` in every property. I would rather called it only on first use, then cached the result and reused it in other properties.